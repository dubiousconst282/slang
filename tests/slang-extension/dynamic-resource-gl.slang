//TEST:SIMPLE(filecheck=CHECK): -stage compute -entry computeMain -target glsl
//TEST:REFLECTION: -stage compute -entry computeMain -target glsl

RWByteAddressBuffer outputBuffer;

// CHECK-DAG: layout(binding = 0, set = 1)
// CHECK-DAG: texture2D g_TextureHeap_1[]

// CHECK-DAG: layout(binding = 0, set = 1)
// CHECK-DAG: utexture2D g_TextureHeap_0[]

[vk::binding(0, 1)]
__DynamicResource<__DynamicResourceKind.General> g_TextureHeap[];

// CHECK-DAG: layout(binding = 2, set = 1)
// CHECK-DAG: sampler g_SamplerHeap_0[]
[vk::binding(2, 1)]
__DynamicResource<__DynamicResourceKind.Sampler> g_SamplerHeap[];

// CHECK-DAG: layout(binding = 0, set = 2)
// CHECK-DAG: texture2D g_SingleDynamicResource_0;
[vk::binding(0, 2)]
__DynamicResource g_SingleDynamicResource;


// CHECK: IndirectAccess_0
// CHECK: g_TextureHeap_0[_S1]
[noinline]
float4 IndirectAccess(__DynamicResource resource, int2 pos)
{
    let texture = resource.as<Texture2D<uint4>>();
    return texture[pos] / 255.0;
}

// CHECK: main
// CHECK-DAG: g_TextureHeap_1[3]
// CHECK-DAG: g_SamplerHeap_0[2]
[numthreads(16)]
void computeMain(int2 tid: SV_DispatchThreadID)
{
    Texture2D texture = g_TextureHeap[3];
    SamplerState sampler = g_SamplerHeap[2].as<SamplerState>();
    float4 value = texture.SampleLevel(sampler, tid / 64.0, 0.0);

    value += IndirectAccess(g_TextureHeap[1], tid);

    value += g_SingleDynamicResource.as<Texture2D>()[tid];

    outputBuffer.Store(tid.x + tid.y * 64, value);
}