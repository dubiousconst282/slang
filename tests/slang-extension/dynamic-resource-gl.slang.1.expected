result code = 0
standard error = {
}
standard output = {
{
    "parameters": [
        {
            "name": "outputBuffer",
            "binding": {"kind": "descriptorTableSlot", "index": 0},
            "type": {
                "kind": "resource",
                "baseShape": "byteAddressBuffer",
                "access": "readWrite"
            }
        },
        {
            "name": "g_TextureHeap",
            "binding": {"kind": "descriptorTableSlot", "space": 1, "index": 0},
            "type": {
                "kind": "array",
                "elementCount": 0,
                "elementType": {
                    "kind": "DynamicResource"
                }
            }
        },
        {
            "name": "g_SamplerHeap",
            "binding": {"kind": "descriptorTableSlot", "space": 1, "index": 2},
            "type": {
                "kind": "array",
                "elementCount": 0,
                "elementType": {
                    "kind": "DynamicResource"
                }
            }
        },
        {
            "name": "g_SingleDynamicResourceA",
            "binding": {"kind": "descriptorTableSlot", "space": 2, "index": 0},
            "type": {
                "kind": "DynamicResource"
            }
        },
        {
            "name": "g_SingleDynamicResourceB",
            "binding": {"kind": "descriptorTableSlot", "space": 2, "index": 1},
            "type": {
                "kind": "DynamicResource"
            }
        }
    ],
    "entryPoints": [
        {
            "name": "computeMain",
            "stage:": "compute",
            "parameters": [
                {
                    "name": "tid",
                    "semanticName": "SV_DISPATCHTHREADID",
                    "type": {
                        "kind": "vector",
                        "elementCount": 2,
                        "elementType": {
                            "kind": "scalar",
                            "scalarType": "int32"
                        }
                    }
                }
            ],
            "threadGroupSize": [16, 1, 1],
            "bindings": [
                {
                    "name": "outputBuffer",
                    "binding": {"kind": "descriptorTableSlot", "index": 0}
                },
                {
                    "name": "g_TextureHeap",
                    "binding": {"kind": "descriptorTableSlot", "space": 1, "index": 0}
                },
                {
                    "name": "g_SamplerHeap",
                    "binding": {"kind": "descriptorTableSlot", "space": 1, "index": 2}
                },
                {
                    "name": "g_SingleDynamicResourceA",
                    "binding": {"kind": "descriptorTableSlot", "space": 2, "index": 0}
                },
                {
                    "name": "g_SingleDynamicResourceB",
                    "binding": {"kind": "descriptorTableSlot", "space": 2, "index": 1}
                }
            ]
        }
    ]
}
}
